### ISSUE TITLE
[BUG] Caching results in wrong QNode return values if only return type differs 

### ISSUE THREAD

**Issue by cvjjm** [PennyLaneAI/pennylane#1706]
### Expected behavior

Even with device caching switched on, computing expectation values and variances of identical circuits should yield correct results.
```
var and expval should be different also with caching: -2.6224875809703296 0.011605015096938232
var and expval should be different also with caching: -2.6224875809703296 0.011605015096938232
```

### Actual behavior

If QNodes that only differ in the return type are executed with caching on, the second QNode returns the same results as the first.
```
var and expval should be different also with caching: -2.6224875809703296 0.011605015096938232
var and expval should be different also with caching: -2.6224875809703296 -2.6224875809703296
```

### Additional information

This is probably not the most minimal example, but hopefully minimal enough to be useful.

### Source code

```shell
import pennylane as qml
import numpy as np

nwires = 2
wires = range(nwires)

hamiltonian = np.array(
    [[-2.5623+0.j    ,  0.    +0.j    ,  0.    +0.j    , 0.    -0.1234j],
     [ 0.    +0.j    , -2.5623+0.j    ,  0.    +0.1234j, 0.    +0.j    ],
     [ 0.    +0.j    ,  0.    -0.1234j, -2.5623+0.j    , 0.    +0.j    ],
     [ 0.    +0.1234j,  0.    +0.j    ,  0.    +0.j    ,-2.5623+0.j    ]])

np.random.seed(172)
params = np.random.randn(2, nwires)
devs = [
    qml.device('default.qubit', wires=nwires),
    qml.device('default.qubit', cache=1, wires=nwires),
]

for dev in devs:

    @qml.qnode(dev, diff_method='parameter-shift')
    def expval_circuit(params):
        qml.templates.BasicEntanglerLayers(params, wires=wires, rotation=qml.RX)
        return qml.expval(qml.Hermitian(hamiltonian, wires=wires))

    @qml.qnode(dev, diff_method='parameter-shift')
    def var_circuit(params):
        qml.templates.BasicEntanglerLayers(params, wires=wires, rotation=qml.RX)
        return qml.var(qml.Hermitian(hamiltonian, wires=wires))

    print("var and expval should be different also with caching:", expval_circuit(params), var_circuit(params))
```

### System information

```shell
Name: PennyLane
Version: 0.19.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/cvjjm/src/covqcstack/qcware/pennylane
Requires: numpy, scipy, networkx, autograd, toml, appdirs, semantic-version, autoray, cachetools, pennylane-lightning
Required-by: covvqetools, pytket-pennylane, PennyLane-Qchem, PennyLane-Lightning
Platform info:           Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-glibc2.10
Python version:          3.8.8
Numpy version:           1.20.1
Scipy version:           1.6.1
Installed devices:
- cov.caching.qubit (covvqetools-0.1.1)
- cov.caching.qulacs (covvqetools-0.1.1)
- cov.parsec.casbox (covvqetools-0.1.1)
- cov.parsec.qubit (covvqetools-0.1.1)
- cov.qubit (covvqetools-0.1.1)
- pytket.pytketdevice (pytket-pennylane-0.1.0)
- lightning.qubit (PennyLane-Lightning-0.18.0)
- default.gaussian (PennyLane-0.19.0.dev0)
- default.mixed (PennyLane-0.19.0.dev0)
- default.qubit (PennyLane-0.19.0.dev0)
- default.qubit.autograd (PennyLane-0.19.0.dev0)
- default.qubit.jax (PennyLane-0.19.0.dev0)
- default.qubit.tf (PennyLane-0.19.0.dev0)
- default.qubit.torch (PennyLane-0.19.0.dev0)
- default.tensor (PennyLane-0.19.0.dev0)
- default.tensor.tf (PennyLane-0.19.0.dev0)

Compilation finished at Thu Sep 30 09:02:29
```

### 

- [X] I have searched exisisting GitHub issues to make sure the issue does not already exist.

**Comment by josh146** [comment]
@cvjjm thanks for reporting this 🐛 !

We should definitely investigate and fix this, however I just want to jump in and quickly advertise an overhaul of how caching works in PennyLane, to get your thoughts and comments.

In the `master` branch of PL, the QNode is currently undergoing an upgrade in order to support:

- Arbitrary nth order derivatives of all operations on hardware
- Custom gradient transforms as the differentiation method
- Internally, if multiple circuits are generated for execution simultaneously, they will be packaged into a single job for execution on the device.

This improved QNode is currently available for testing as [`qml.beta.qnode`](pennylane.rtfd.io/en/latest/code/api/pennylane.beta.qnode.html), and will be available in the next release.

As part of this improvement, caching has _also_ been improved, and has been made device independent. You can now create your own cache object (it can be simply a dictionary, or a `cachetools.LRUCache`, for example), and pass it to QNodes you want to use it in:

```python
dev = qml.device('default.qubit', wires=nwires)

cache = {}

@qml.beta.qnode(dev, diff_method='parameter-shift', cache=cache)
def expval_circuit(params):
    qml.templates.BasicEntanglerLayers(params, wires=wires, rotation=qml.RX)
    return qml.expval(qml.Hermitian(hamiltonian, wires=wires))

@qml.beta.qnode(dev, diff_method='parameter-shift', cache=cache)
def var_circuit(params):
    qml.templates.BasicEntanglerLayers(params, wires=wires, rotation=qml.RX)
    return qml.var(qml.Hermitian(hamiltonian, wires=wires))
```
```pycon
>>> print(expval_circuit(params), var_circuit(params)))
-2.6224875809703287, 0.011605015096940896
>>> print(cache)
{-2104543859499562169: array([-2.62248758]), 3268478774057874589: array([0.01160502])}
>>> expval_circuit(params), var_circuit(params)) # evaluate circuits again
>>> dev.num_executions
2
```

If this is something that will fit your caching workflow, or if you have any thoughts, please let me know!

**Comment by cvjjm** [comment]
Very nice! Yes, that kind of caching looks very useful.

How does it play together with the new batch transforms?

**Comment by josh146** [comment]
> How does it play together with the new batch transforms?

This is a good question. I'm not sure if it works currently out of the box, _but_ [`qml.beta.qnode`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.beta.qnode.html) and the batch transforms share the same execution pipeline: [`qml.execute()`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.execute.html).

As a result, it should be fairly trivial to set it up so that a batch transform can share a cache across evaluations 🙂

**Comment by josh146** [comment]
@cvjjm, it just required a small tweak in PR #1708, but now caching also works with batch transforms:

```python
dev = qml.device("default.qubit", wires=1)

cache = {}

@qml.beta.qnode(dev, cache=cache)
def circuit(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

@qml.batch_transform
def transform_fn(tape):
    return [tape.copy(), tape.copy()], None

fn = transform_fn(circuit)
```

```pycon
>>> fn(0.5)
[tensor([0.87758256], requires_grad=True), tensor([0.87758256], requires_grad=True)]
>>> fn(0.5)
[tensor([0.87758256], requires_grad=True), tensor([0.87758256], requires_grad=True)]
>>> cache
{1933081548548453010: tensor([0.87758256], requires_grad=True)}
>>> circuit.device.num_executions
1
```
